src/main/java/uk/gov/hmcts/ccd/domain/service/getevents/AuthorisedGetEventsOperation.java

```diff
@@ -74,12 +85,19 @@ public AuthorisedGetEventsOperation(@Qualifier("classified") GetEventsOperation
             throw new ValidationException("Cannot find case type definition for  " + caseTypeId);
         }
 
-        Set<String> userRoles = userRepository.getUserRoles();
-        if (userRoles == null || userRoles.isEmpty()) {
+        if (caseUserRoles == null || caseUserRoles.isEmpty()) {
             throw new ValidationException("Cannot find user roles for the user");
         }
 
-        return verifyReadAccess(events, userRoles, caseType);
+        return verifyReadAccess(events, caseUserRoles, caseType);
+    }
+
+    private Set<String> getUserRoles(String caseId) {
+        return Sets.union(userRepository.getUserRoles(),
+            caseUserRepository
```

Comment:
Important: I've noticed that CaseUserRepository does not have a Cached version. Compare with UserRepository. This means every call to retrieve the case roles done **as part of the same endpoint call** will reach the db. Can we introduce a CachedCaseUserRepository?

Reply:
Will look into this....

Reply:
CachedCaseUserRepository has been introduced... Please review...

Reply:
👍